app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingEvidenceController.scala

```diff
@@ -0,0 +1,1012 @@
+                        _.copy(journeyStatus = Some(fillingOutReturn.copy(draftReturn = newDraftReturn)))
+                      )
+                    )
+
+              } yield reference
+
+              result.fold(
+                error => {
+                  logger.warn(s"failed to upload file with error: $error")
+                  errorHandler.errorResult(None)
+                },
+                ref => Redirect(routes.UploadSupportingEvidenceController.uploadSupportingEvidenceVirusCheck(ref))
+              )
+            }
+          )
+          .getOrElse {
+            logger.warn("missing file key")
+            Future.successful(errorHandler.errorResult(None))
+          }
+      }
+    }
+
+  def uploadSupportingEvidenceVirusCheck(reference: String): Action[AnyContent] =
+    authenticatedActionWithSessionData.async { implicit request =>
+      withUploadSupportingEvidenceAnswers(request) { (draftReturnId, _, _, _, _) =>
+        val result = for {
+          fd <- upscanService
+                 .getUpscanFileDescriptor(DraftReturnId(draftReturnId.toString), UpscanInitiateReference(reference))
+        } yield fd
+
+        result.fold(
+          error => {
+            logger.warn(s"failed to get file descriptor information: $error")
+            errorHandler.errorResult(None)
+          }, {
+            case Some(upscanFileDescriptor) =>
+              if (upscanFileDescriptor.status === UPLOADED | upscanFileDescriptor.status === FAILED) {
+                Ok(
+                  uploadSupportingEvidenceUpscanCheckPage(
+                    UpscanInitiateReference(reference),
+                    upscanFileDescriptor.status
+                  )
+                )
+              } else {
+                Redirect(routes.UploadSupportingEvidenceController.checkYourAnswers())
+              }
+            case None => errorHandler.errorResult(None)
+          }
+        )
+      }
+    }
+
+  def uploadSupportingEvidenceVirusCheckSubmit(reference: String): Action[AnyContent] =
+    authenticatedActionWithSessionData.async { implicit request =>
+      Redirect(routes.UploadSupportingEvidenceController.uploadSupportingEvidenceVirusCheck(reference))
+    }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Var", "org.wartremover.warts.Any"))
+  private def handleGetFileDescriptorResult(
+    multipart: MultipartFormData[Files.TemporaryFile],
+    upscanFileDescriptor: UpscanFileDescriptor
+  ): Either[Error, MultipartFormData[Source[ByteString, Any]]] = {
+    val userFile =
+      multipart.files
+        .map(file => file.copy(ref = FileIO.fromPath(file.ref.path): Source[ByteString, Any]))
+    val prepared: MultipartFormData[Source[ByteString, Any]] =
+      multipart
+        .copy(
+          files = userFile,
+          dataParts = upscanFileDescriptor.fileDescriptor.uploadRequest.fields
+            .mapValues(fieldValue => Seq(fieldValue))
+        )
+    Right(prepared)
```

Comment:
you're always returning a `Right` here - the return type of this function would be better without the `Either`

Reply:
will be addressed after 6th April